go-controller/pkg/ovn/policy.go

```diff
@@ -1270,93 +1336,35 @@ func (oc *Controller) createNetworkPolicy(policy *knet.NetworkPolicy, aclLogging
 					podSelector:       toJSON.PodSelector,
 				})
 			}
-			np.egressPolicies = append(np.egressPolicies, egress)
 		}
-		np.Unlock()
 
-		for _, handler := range policyHandlers {
-			var err error
-			if handler.namespaceSelector != nil && handler.podSelector != nil {
-				// For each rule that contains both peer namespace selector and
-				// peer pod selector, we create a watcher for each matching namespace
-				// that populates the addressSet
-				err = oc.addPeerNamespaceAndPodHandler(handler.namespaceSelector, handler.podSelector, handler.gress, np)
-			} else if handler.namespaceSelector != nil {
-				// For each peer namespace selector, we create a watcher that
-				// populates ingress.peerAddressSets
-				err = oc.addPeerNamespaceHandler(handler.namespaceSelector, handler.gress, np)
-			} else if handler.podSelector != nil {
-				// For each peer pod selector, we create a watcher that
-				// populates the addressSet
-				err = oc.addPeerPodHandler(policy, handler.podSelector,
-					handler.gress, np)
-			}
-			if err != nil {
-				return fmt.Errorf("failed to handle policy handler selector: %v", err)
-			}
+		// 3. Add policy to default deny port group
```

Comment:
yeah that makes sense, I definitely need to add an explanation of this order, since it wasn't obvious before
But also we know that if peer object handler fails, and needs to be retried, we still can block connection before it was allowed, but that should happen rarely I guess.

One more thing: while looking at networkpolicy test failures, I see the following:
networkpolicy is created for existing pods, right after that first connectivity check is run. If connection should be denied, it can be allowed and cause the test failure, because pod wasn't added to default deny port group yet (because peer handlers take some time). In general I guess that's ok to assume that network policy creation takes some time and prefer saving existing connections and delaying deny rules creation, but k8s tests don't think so :)

Reply:
yeah this has been a struggle with the tests. That's why we wanted to use a status field in the network policy spec to indicate when the policy was enforcing from the CNI perspective:
https://github.com/kubernetes/api/blob/master/networking/v1/types.go#L237

But I think there is resistance to adding checks for that status in e2e tests. Maybe @astoycos can comment more

Reply:
Yeah there's major concerns around actually being able to specify when a policy is enforcing... as shown in https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2943-networkpolicy-status

and 

https://github.com/kubernetes/enhancements/pull/2137

For example a major edge case is create a pod and a NetworkPolicy, and then change the pod labels; how do you know when the NP applies to it? 

But I've been thinking alot about this lately.... If we _think_ it's possible to implement such a "enforcing Status" in Ovn-Kubernetes, I _could_ add a toggle in our Downstream k8s fork to check for that condition before running tests..... Now this is a bit murky because playing with upstream "conformance" tests in our downstream fork might raise some eyebrows

Reply:
Well I guess this is upstream :) but IF Ovn-K could actually adequately implement such a status that would be grounds to add an upstream "test config"  for the "only run test when we're ready" functionality.

Reply:
Your changes for the ordering look good

Reply:
@astoycos re: "For example a major edge case is create a pod and a NetworkPolicy, and then change the pod labels; how do you know when the NP applies to it?"

I think the status in the network policy would only reflect that the network policy has been programmed by the controller and upon network policy creation/update. Pod changes would not mess with that status.

This solves the problem in the tests where pods are already created, then a network policy is created in the API, and immediately connections are attempted.

For ovnk we have discussed going with a CRD where we can indicate to our CNI to delay wiring a pod until network policy has been evaluated for that pod.